engine/grid.py

```diff
@@ -300,73 +307,164 @@ def find_border_nodes(self):
             'border_nodes' will be initialized.
         """
         border_list = []
-        cols = self.nodes.shape[1]
-        rows = self.nodes.shape[0]
         leftmost_node = None
+        rightmost_node = None
         leftmost_node_pos = None
-        for row in range(rows):
-            for col in range(cols):
+        rightmost_node_pos = None
+        
+        for row in range(self.num_rows):
+            for col in range(self.num_cols):
                 node = self.nodes[row][col]
-                if node.is_active and self.is_on_border(row, col, rows-1, cols-1):
+                #border node 1 left active node 
+                if node.is_active and self.is_on_border(row, col, self.num_rows-1, self.num_cols-1):
                     # check if this is an active node and on the border
                     self.nodes[row][col].is_border = True
                     border_list.append((node, row, col))
                     if leftmost_node_pos is None or col < leftmost_node_pos[1]:
                         leftmost_node = node
                         leftmost_node_pos = (row, col)
+                if node.is_active and self.is_on_border(row, col, self.num_rows , self.num_cols):
+                    # check if this is an active node and on the border
+                    self.nodes[row][col].is_border = True
+                    if rightmost_node_pos is None or row > rightmost_node_pos[0]:
```

Comment:
when updating rightmost_node_pos and rightmost_node, why are we concerned about the row and not the col?

Reply:
It seems like my code is different, I have this
`  def find_border_nodes(self):
        """
            Find all activated border nodes on the grid.

            This function loops through all the nodes, checks if a particular node has
            been activated, and if so checks to see if that node is a border node. At the
            end of the function call, fields 'leftmost_node', 'leftmost_node_pos', and
            'border_nodes' will be initialized.
        """
        border_list = []
        leftmost_node = None
        rightmost_node = None
        leftmost_node_pos = None
        rightmost_node_pos = None
        
        for row in range(self.num_rows):
            for col in range(self.num_cols):
                node = self.nodes[row][col]
                if node.is_active and self.is_on_border(row, col, self.num_rows-1, self.num_cols-1):
                    # check if this is an active node and on the border
                    self.nodes[row][col].is_border = True
                    border_list.append((node, row, col))
                    if leftmost_node_pos is None or col < leftmost_node_pos[1]:
                        leftmost_node = node
                        leftmost_node_pos = (row, col)
                    
                    
                        
        self.border_nodes = border_list
        self.leftmost_node = leftmost_node
        self.leftmost_node_pos = leftmost_node_pos
        self.rightmost_node = rightmost_node
        self.rightmost_node_pos = rightmost_node_pos`

Reply:
However, the rightmost_node and rightmost_node_pos are never redefined and never used later. I think I added it during the zigzag traversal, but since we always traverse starting from the left side now, I can remove those attributes from the class

Reply:
ok, resolve when finished